python/src/eval.cc

```diff
@@ -0,0 +1,61 @@
+#include "internal/eval.hh"
+#include "internal/errors.hh"
+#include "internal/nix-to-python.hh"
+#include "internal/python-to-nix.hh"
+#include <Python.h>
+
+#include <cxxabi.h>
+#include <store-api.hh>
+
+namespace nix::python {
+
+const char * currentExceptionTypeName()
+{
+    int status;
+    auto res = abi::__cxa_demangle(abi::__cxa_current_exception_type()->name(), 0, 0, &status);
+    return res ? res : "(null)";
+}
+
+static PyObject * _eval(const char * expression, PyObject * vars)
+{
+    nix::Strings storePath;
+    nix::EvalState state(storePath, nix::openStore());
+
+    nix::Env * env = nullptr;
+    auto staticEnv = pythonToNixEnv(state, vars, &env);
+    if (!staticEnv) {
+        return nullptr;
+    }
+    auto staticEnvPointer = std::make_shared<nix::StaticEnv>(*staticEnv);
+
+    auto e = state.parseExprFromString(expression, ".", staticEnvPointer);
+    nix::Value v;
+    e->eval(state, *env, v);
+
+    state.forceValueDeep(v);
```

Comment:
should we release the GIL here? (Py_BEGIN_ALLOW_THREADS)

Reply:
Didn't know what that was originally, but reading about it, yes we should do this here. I added a commit for this.

Reply:
Turns out this causes a segfault, though I don't really know why: https://github.com/NixOS/nix/pull/7735/commits/e11273f8c982e76111da604a453f4759c80a13d4

Reply:
`forceValueDeep` might throw an exception, which would unwind out of this function without reacquiring the GIL.

Reply:
Ahh that would make sense, I'll try your suggestion out

Reply:
Yup that worked!